adapters/mobilefuse/mobilefuse.go

```diff
@@ -0,0 +1,193 @@
+package mobilefuse
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"net/http"
+	"strconv"
+	"text/template"
+)
+
+type MobilefuseAdapter struct {
+	EndpointTemplate template.Template
+}
+
+func NewMobilefuseBidder(endpointTemplate string) adapters.Bidder {
+	parsedTemplate, errs := template.New("endpointTemplate").Parse(endpointTemplate)
+
+	if errs != nil {
+		glog.Fatal("Unable parse endpoint template: " + errs.Error())
+		return nil
+	}
+
+	return &MobilefuseAdapter{EndpointTemplate: *parsedTemplate}
+}
+
+func (adapter *MobilefuseAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var adapterRequests []*adapters.RequestData
+
+	adapterRequest, errs := adapter.makeRequest(request)
+
+	if errs == nil {
+		adapterRequests = append(adapterRequests, adapterRequest)
+	}
+
+	return adapterRequests, errs
+}
+
+func (adapter *MobilefuseAdapter) MakeBids(incomingRequest *openrtb.BidRequest, outgoingRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d.", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d.", response.StatusCode),
+		}}
+	}
+
+	var incomingBidResponse openrtb.BidResponse
+
+	if err := json.Unmarshal(response.Body, &incomingBidResponse); err != nil {
+		return nil, []error{err}
+	}
+
+	outgoingBidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
+
+	for _, seatbid := range incomingBidResponse.SeatBid {
+		for i := range seatbid.Bid {
+			outgoingBidResponse.Bids = append(outgoingBidResponse.Bids, &adapters.TypedBid{
+				Bid:     &seatbid.Bid[i],
+				BidType: adapter.getBidType(seatbid.Bid[i].ImpID, incomingRequest.Imp),
+			})
+		}
+	}
+
+	return outgoingBidResponse, nil
+}
+
+func (adapter *MobilefuseAdapter) makeRequest(bidRequest *openrtb.BidRequest) (*adapters.RequestData, []error) {
+	var errs []error
+
+	mobilefuseBidRequest := *bidRequest
```

Comment:
Small optimization: create the `openrtb.BidRequest` copy only if valid imps are found. Given that `imp` is a safe copy in the context of the `for` statement in line 170, a small optimization might be to create the expensive `mobilefuseBidRequest := *bidRequest` only when needed. It could look like:
```
 81   func (adapter *MobilefuseAdapter) makeRequest(bidRequest *openrtb.BidRequest) (*adapters.RequestData, []error) {
 82       var errs []error
 83  
 84 -     mobilefuseBidRequest := *bidRequest
 85  
 86       mobilefuseExtension, errs := adapter.getMobilefuseExtension(bidRequest)
 87  
 88   *--  9 lines: if mobilefuseExtension == nil {---------------------------------------------------------------------
 97  
 98 -     adapter.modifyBidRequest(&mobilefuseBidRequest, mobilefuseExtension)
    +     validImps := getValidImps(bidRequest, mobilefuseExtension)
 99  
100 -     if len(mobilefuseBidRequest.Imp) == 0 {
    +     if len(validImps) == 0 {
101           err := fmt.Errorf("No valid imps")
102           errs = append(errs, err)
103           return nil, errs
104       }
    +     mobilefuseBidRequest := *bidRequest
    +     mobilefuseBidRequest.Imp = validImps
105  
106       body, err := json.Marshal(mobilefuseBidRequest)
107   *-- 15 lines: if err != nil {-------------------------------------------------------------------------------------
122   }
123
124 *-- 42 lines: func (adapter *MobilefuseAdapter) getMobilefuseExtension(request *openrtb.BidRequest) (*openrtb_ext.ExtImpMobilefuse, []error) {--
166
167 - func (adapter *MobilefuseAdapter) modifyBidRequest(bidRequest *openrtb.BidRequest, ext *openrtb_ext.ExtImpMobilefuse) {
    + func (adapter *MobilefuseAdapter) getValidImps(bidRequest *openrtb.BidRequest, ext *openrtb_ext.ExtImpMobilefuse) []openrtb.Imp {
168       var validImps []openrtb.Imp
169  
170       for _, imp := range bidRequest.Imp {
171           if imp.Banner != nil || imp.Video != nil {
172               if imp.Banner != nil && imp.Video != nil {
173                   imp.Video = nil
174               }
175  
176               imp.TagID = strconv.Itoa(ext.PlacementId) //`imp` is a copy of `bidRequest.Imp[i]`
177               imp.Ext = nil
178               validImps = append(validImps, imp)
179           }
180  
181           break
182       }
183  
184 -     bidRequest.Imp = validImps
    +     return validImps
185   }
adapters/mobilefuse/mobilefuse.go
```

Reply:
Makes sense, I've implemented. All for micro-optimizations when billions of executions are involved. :)

Reply:
Thanks